lib/core/unittest/CConcurrencyTest.cc

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+#include "CConcurrencyTest.h"
+
+#include <core/CLogger.h>
+#include <core/Concurrency.h>
+
+#include <numeric>
+
+using namespace ml;
+
+void CConcurrencyTest::testAsyncWithExecutors() {
+
+    core::stopDefaultAsyncExecutor();
+
+    std::string tags[]{"sequential", "parallel"};
+
+    for (std::size_t t = 0; t < 2; ++t) {
+        LOG_DEBUG(<< "Testing " << tags[t]);
+
+        auto result =
+            core::async(core::defaultAsyncExecutor(), []() { return 42; });
+        CPPUNIT_ASSERT_EQUAL(42, result.get());
+
+        result = core::async(core::defaultAsyncExecutor(),
+                             [](int i) { return i; }, 43);
+        CPPUNIT_ASSERT_EQUAL(43, result.get());
+
+        result = core::async(core::defaultAsyncExecutor(),
+                             [](int i, int j) { return i + j; }, 22, 22);
+        CPPUNIT_ASSERT_EQUAL(44, result.get());
+
+        core::startDefaultAsyncExecutor();
+    }
+
+    core::stopDefaultAsyncExecutor();
+}
+
+void CConcurrencyTest::testParallelForEach() {
+
+    core::stopDefaultAsyncExecutor();
+
+    std::vector<int> values(10000);
+
+    LOG_DEBUG(<< "Test empty");
```

Comment:
nit: could be split into separate test cases?

Reply:
[780ebb0](https://github.com/elastic/ml-cpp/pull/322/commits/780ebb0177036c0ba47990fc9cf44283ad7e8406).